### ISSUE TITLE
Standardize on the Python shebang line

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2208]
|<img alt="thomasvs's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| thomasvs reported|
|-|-|
|Trac ID|trac#2208|
|Type|enhancement|
|Created|2006-11-01 14:41:31Z|
|Branch|https://github.com/twisted/twisted/tree/standardize-shebang-2208|

This is an example from Names:
```
W: python-twisted-names doc-file-dependency /usr/share/doc/python-twisted-names-0.3.0/examples/dns-service.py /usr/bin/python
W: python-twisted-names doc-file-dependency /usr/share/doc/python-twisted-names-0.3.0/examples/gethostbyname.py /usr/bin/env

```

I've seen a lot of other different shebangs in other modules as well.  Maybe Twisted should standardize on one ?

Attachments:

* [shebang-2208.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/ab6/ab6ad881e7d1e78559ff2f9cb718f888bcd4d596/d61f21a2f4afc1cb87cb95fde9c8b1bf86d9833d.patch) (15802 bytes) - added by collab on 2008-07-04 00:42:00Z - patch against r24235
* [shebang-2208.2.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/ab6/ab6ad881e7d1e78559ff2f9cb718f888bcd4d596/28ac68b12fab604e7a30e059661ab52f755bf009.patch) (15973 bytes) - added by thijs on 2008-07-12 23:43:41Z - patch against r24262

<details><summary>Searchable metadata</summary>

```
trac-id__2208 2208
type__enhancement enhancement
reporter__thomasvs thomasvs
priority__normal normal
milestone__ 
branch__branches_standardize_shebang_2208 branches/standardize-shebang-2208
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1162392091000000 1162392091000000
changetime__1217872466000000 1217872466000000
version__None None
owner__ 
cc__thijs
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_1"></a>|@glyph set owner to thomasvs|
|-|-|

Yes.  Propose on the mailing list?

**Comment by twisted-trac** [comment]
|<img alt="collab's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_2"></a>|collab commented|
|-|-|

[Wikipedia](http://en.wikipedia.org/wiki/Shebang_(Unix)#Examples) lists ```#!/usr/bin/python``` is the typical line.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun commented|
|-|-|

`#!/usr/bin/python` is what I would go for.

**Comment by twisted-trac** [comment]
|<img alt="collab's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_4"></a>|collab set owner to collab<br>collab set status to `assigned`|
|-|-|

Replying to [exarkun](#note_3):
> `#!/usr/bin/python` is what I would go for.

I'll make a patch for it.

**Comment by twisted-trac** [comment]
|<img alt="collab's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_5"></a>|collab removed owner<br>collab set status to `new`|
|-|-|

Attached a [[shebang-2208.patch](None/shebang-2208.patch) patch] that fixes the problem (looked through all .py files in the trunk).

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun set owner to collab|
|-|-|

We should update `doc/development/policy/coding-standard.xhtml` to include a section about example code, including a note to use `#!/usr/bin/python` for runnable examples, too.

**Comment by twisted-trac** [comment]
|<img alt="collab's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_7"></a>|collab removed owner|
|-|-|

Replying to [exarkun](#note_6):
> We should update `doc/development/policy/coding-standard.xhtml` to include a section about example code, including a note to use `#!/usr/bin/python` for runnable examples, too.

I think adding a section for example code would require it's own ticket, there's nothing in the coding-standard.xhtml yet. There is a however a section for Scripts, which I updated in the attached [[shebang-2208.patch](None/shebang-2208.patch) patch] that also replaced the shebang lines in the various script files.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun set owner to @thijstriemstra|
|-|-|

We could make a different ticket for updating the coding standard, but updating it now would be better.  If we update it now, at the same time as we fix all of the `#!`, then everything in the codebase is correct and the correct way to do things is documented.  If we fix all the `#!` now and update the coding standard later, then between now and later, incorrect `#!` may be introduced which will have to be fixed.  Alternatively, we could fix the coding standard in a separate ticket first and only then resolve this ticket.  However, there's very little code in the attached patch, and these are very related changes, so I don't see the need to create another ticket at all.

The coding standard change doesn't need to be long.  A couple sentences should do.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)|@thijstriemstra set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_12"></a>|@thijstriemstra removed owner<br>@thijstriemstra set status to `new`|
|-|-|

Added a new [[shebang-2208.2.patch](None/shebang-2208.2.patch) patch] for review.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_13"></a>|@exarkun set owner to @thijstriemstra|
|-|-|

`p` tags need to be closed in lore docs.  You can check the .xhtml files for validity with `lore -olint <filename>`.  A couple files have problems, although some of them may be pre-existing (if you can fix those too though, that'd be awesome :).

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_14"></a>|@thijstriemstra set status to `assigned`|
|-|-|

Replying to [exarkun](#note_13):
> `p` tags need to be closed in lore docs.  You can check the .xhtml files for validity with `lore -olint <filename>`.  A couple files have problems, although some of them may be pre-existing (if you can fix those too though, that'd be awesome :).

Alright, I'll update the patch.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_15"></a>|@thijstriemstra commented|
|-|-|

(In [24435]) Branching to 'standardize-shebang-2208'

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_16"></a>|@thijstriemstra removed owner<br>@thijstriemstra set status to `new`|
|-|-|

Replying to [thijs](#note_14):
> Replying to [exarkun](#note_13):
> > `p` tags need to be closed in lore docs.  You can check the .xhtml files for validity with `lore -olint <filename>`. 

I fixed the p tag issue I added, couldn't find any others and your suggested 'find -name '*.xhtml' | xargs lore -olint' threw errors on osx leopard. 

The doc builder didn't report any errors: http://buildbot.twistedmatrix.com/builders/documentation/builds/712

Putting it up for review.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_17"></a>|@exarkun set owner to @thijstriemstra|
|-|-|

osx requires the path argument to find - `find . -name '*.xhtml' | xargs lore -olint` would probably work, but the doc builder is just as good.

The wording in `coding-standard.xhtml` suggests that the code snippet which only includes a `#!` example is actually a `#!` example and the _Twisted running-from-SVN_ header, which it isn't.  I'd move the running-from-SVN part of that sentence to after the example.

The rest looks good.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_18"></a>|@thijstriemstra set owner to @exarkun|
|-|-|

Replying to [exarkun](#note_17):
> osx requires the path argument to find - `find . -name '*.xhtml' | xargs lore -olint` would probably work, but the doc builder is just as good.

That worked but it didn't report any specific errors about p tags, unless I didn't see Lore's error description.

>
> The wording in `coding-standard.xhtml` suggests that the code snippet which only includes a `#!` example is actually a `#!` example and the _Twisted running-from-SVN_ header, which it isn't.  I'd move the running-from-SVN part of that sentence to after the example.

I changed it in r24467, hope I understand you correctly. Putting it up for review.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_19"></a>|@exarkun set owner to @thijstriemstra|
|-|-|

Looks great now, thanks. :)  Please merge.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_20"></a>|@thijstriemstra set status to `closed`|
|-|-|

(In [24516]) Merge standardize-shebang-2208: Standardize on the Python shebang line #!/usr/bin/python

Author: thijs
Reviewer: exarkun
Fixes: [#2208](https://github.com/twisted/twisted/issues/2208)

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
